optaplanner-core/src/test/java/org/optaplanner/core/config/solver/EnvironmentModeTest.java

```diff
@@ -0,0 +1,251 @@
+        ((DefaultSolver<TestdataSolution>) solver2).addPhaseLifecycleListener(listener2);
+
+        try (AutoCloseableSoftAssertions softly = new AutoCloseableSoftAssertions()) {
+            for (int i = 0; i < NUMBER_OF_TIMES_RUN; i++) {
+                solver1.solve(inputProblem);
+                solver2.solve(inputProblem);
+                softly.assertThat(listener.getScores()).as("Score steps should be the same "
+                                                                   + "in a reproducible environment mode.")
+                        .isEqualTo(listener2.getScores());
+            }
+        }
+    }
+
+    private void assertDifferentScoreSeries(Solver<TestdataSolution> solver1, Solver<TestdataSolution> solver2) {
+        TestdataStepScoreListener listener = new TestdataStepScoreListener();
+        TestdataStepScoreListener listener2 = new TestdataStepScoreListener();
+
+        ((DefaultSolver<TestdataSolution>) solver1).addPhaseLifecycleListener(listener);
+        ((DefaultSolver<TestdataSolution>) solver2).addPhaseLifecycleListener(listener2);
+
+        try (AutoCloseableSoftAssertions softly = new AutoCloseableSoftAssertions()) {
+            for (int i = 0; i < NUMBER_OF_TIMES_RUN; i++) {
+                solver1.solve(inputProblem);
+                solver2.solve(inputProblem);
+                softly.assertThat(listener.getScores()).as("Score steps should not be the same in a non-reproducible environment mode. "
+                                                                   + "This might be possible because searchSpace is not infinite and "
+                                                                   + "two different random scenarios can have the same results. "
+                                                                   + "Run test again.")
+                        .isNotEqualTo(listener2.getScores());
+            }
+        }
+    }
+
+    private void assertGeneratingSameNumbers(RandomFactory factory1, RandomFactory factory2) {
+        Random random = factory1.createRandom();
+        Random random2 = factory2.createRandom();
+
+        for (int i = 0; i < EnvironmentModeTest.NUMBER_OF_RANDOM_NUMBERS_GENERATED; i++) {
+            Assertions.assertThat(random.nextInt()).as("Random factories should generate the same results "
+                                                               + "in a reproducible environment mode.")
+                    .isEqualTo(random2.nextInt());
+        }
+    }
+
+    private void assertGeneratingDifferentNumbers(RandomFactory factory1, RandomFactory factory2) {
+        Random random = factory1.createRandom();
+        Random random2 = factory2.createRandom();
+
+        for (int i = 0; i < EnvironmentModeTest.NUMBER_OF_RANDOM_NUMBERS_GENERATED; i++) {
+            Assertions.assertThat(random.nextInt()).as("Random factories should not generate exactly the same results "
```

Comment:
Check the `SoftAssertions` pattern. You will be able to get the entire loop covered by one assertion.